scalding-quotation/src/test/scala/com/twitter/scalding/quotation/ProjectionTest.scala

```diff
@@ -0,0 +1,168 @@
+package com.twitter.scalding.quotation
+
+import org.scalatest.MustMatchers
+import org.scalatest.FreeSpec
+
+trait S
+
+trait T1 extends S
+trait T2
+
+trait P1 extends S
+trait P2
+
+class ProjectionTest extends Test {
+
+  val t1 = TypeReference(typeName[T1])
+  val p1 = Property(t1, Accessor("p1"), typeName[P1])
+
+  val t2 = TypeReference(TypeName(classOf[T2].getName))
+  val p2 = Property(t2, Accessor("p2"), typeName[P2])
+
+  "Projection" - {
+    "andThen" - {
+      "TypeReference" in {
+        t1.andThen(p1.accessor, p1.typeName) mustEqual p1
+      }
+      "Property" in {
+        p1.andThen(Accessor("p2"), TypeName("p2t")) mustEqual
+          Property(p1, Accessor("p2"), TypeName("p2t"))
+      }
+    }
+
+    "toString" - {
+      "TypeReference" - {
+        "simple" in {
+          t1.toString mustEqual "T1"
+        }
+        "ignores package" in {
+          TypeReference(TypeName("com.twitter.Test1")).toString mustEqual "Test1"
+        }
+      }
+      "Property" in {
+        p1.toString() mustEqual "T1.p1"
+      }
+    }
+  }
+
+  "Projections" - {
+    "empty" in {
+      Projections.empty.set mustEqual Set()
+    }
+    "apply" - {
+      "simple" in {
+        val set = Set[Projection](p1)
+        Projections(set).set mustEqual set
+      }
+      "paths merge" - {
+        "simple" in {
+          val set = Set[Projection](p1, t1)
+          Projections(set).set mustEqual Set(t1)
+        }
+        "nested" in {
+          val px = p1.andThen(Accessor("x"), TypeName("X"))
+          val set = Set[Projection](px, t1)
+          Projections(set).set mustEqual Set(t1)
+        }
+      }
+    }
+    "flatten" - {
+      "empty" in {
+        Projections.flatten(Nil).set mustEqual Set()
+      }
+      "non-empty" in {
+        val list = List(
+          Projections(Set(p1)),
+          Projections(Set(p2)))
+        Projections.flatten(list).set mustEqual Set(p1, p2)
+      }
+      "non-empty with merge" in {
+        val list = List(
+          Projections(Set(t1)),
+          Projections(Set(p1)))
+        Projections.flatten(list).set mustEqual Set(t1)
+      }
+    }
+
+    "++" - {
+      "simple" in {
+        val p = Projections(Set(p1)) ++ Projections(Set(p2))
+        p.set mustEqual Set(p1, p2)
+      }
+      "with merge" in {
```

Comment:
seems like we could do a scalacheck test here. We could test that for all `List[Projections]` then if we `++` all of them, we have a final result where there are not Projections which are "suffixes" of another.

Reply:
I'm not a big fan of property-based testing. Is there something that would have better coverage if I use it?

Reply:
WHAT!?!? Who is not a big fan of property based checks!? Why don't you like it? I love it. It gives you an ability to clearly state properties and a decent approach at checking if what you claim is true actually is. Of course, with poor generators, it often doesn't hit tricky cases, but I have found it to be hugely helpful.

Reply:
If the code is refactored into simple functions that don't require complex structures, unit tests are easier to implement and reason about. Property-based tests take longer to implement, are much harder to debug, and don't bring many benefits if compared to simple unit tests.

Reply:
I guess the reason I brought it up is that I thought I noticed a bug which I think a scalacheck would have caught.

Reply:
Sure, if I had thought about that scenario and checked the property :) The same is valid for unit tests. I can implement the property-based test if you feel strongly about it, though.